Added application classes for qiskit-finance

a-matsuo opened the pull request:
### Summary
This PR is for application classes in the Qiskit finance. 

### Details and comments
New `Portfolio` and `PortfolioDiversification` classes inherit `OptimizationApplication` in the Qiskit optimization and they work like other application classes such as `Maxcut`. `BaseApplication` and other application classes were introduced in https://github.com/Qiskit/qiskit-optimization/pull/31

- [x] Portfolio([example](https://github.com/a-matsuo/optimization_class/blob/main/examples/portfolio_example.ipynb))
- [x] PortfolioDiversification

Also, `EstimationApplication` is introduced for other applications which uses the amplitude estimation.
- [x] EuropeanCallExpectedValue
- [x] EuropeanCallDelta
- [x] FixedIncomeExpectedValue

Question
- [x] Naming: 2 EuropeanCallExpectedValue classes exist in the circuit library and the application folder
- [x] Do we need `ising` and `estimation` folders under `applications`?
- [x] How to pass uncertainty_model information? Currently, the corresponding quantum circuit is passed. Do we want to pass it as a dictionary?

stefan-woerner commented:
Hi @a-matsuo, on your questions:
- let's rename to rename the folders to: `applications/optimization` (instead of `ising`) and `applications/estimation`
- please use circuits for uncertainty models for now (we can discuss in the future to extend)
- you can name the application `EuropeanCallPricing` and the circuit `EuropeanCallPricingObjective`. For the `Delta` you can use `EuropeanCallDelta` (for the app) and `EuropeanCallDeltaObjective` for the circuit, so that naming is consistent.

a-matsuo commented:
Hi @stefan-woerner, thank you for answering my questions. 
I finished implementing `PortfolioDiversification` as well. I'm going to add such as unittests and comments now.

woodsp-ibm commented on qiskit_finance/applications/estimation/estimation_application.py:
```diff
@@ -10,16 +10,20 @@
+from qiskit.algorithms.amplitude_estimators import EstimationProblem
+
 
 class EstimationApplication(ABC):
```
Do we expect to have an `interpret` method in here similar to OptimizationApplication

a-matsuo replied:
Honestly, somehow, I didn't even think about it, but you're right. It's better and natural to have an `interpret` method.

a-matsuo commented:
I added unittests and updated tutorials to work with new application classes. This PR is ready now. Could you please review the PR? @stefan-woerner @woodsp-ibm

woodsp-ibm commented on qiskit_finance/applications/estimation/estimation_application.py:
```diff
@@ -0,0 +1,40 @@
+    @abstractmethod
+    def interpret(self, result: AmplitudeEstimatorResult) -> float:
+        """Convert the calculation result of the problem
+        (:class:`~qiskit.algorithms.amplitude_estimators.AmplitudeEstimatorResult`)
```
I do not know whether we will be able to hyperlink across like this (i.e the ref to the class in the other repo) as I believe the intent is to build out the docs from each repo independently.

woodsp-ibm commented on qiskit_finance/applications/estimation/fixed_income_pricing.py:
```diff
@@ -0,0 +1,92 @@
+            initial_interests: The initial interest rates / offsets for the interest rates.
+            cash_flow: The cash flow time series.
+            rescaling_factor: The scaling factor used in the Taylor approximation.
+            bounds: The bounds for return values the assets can attain.
```
What is the tuple here as part of the bound? max, min or min, max. Maybe there is some standard for this and it conforms to that. It might nevertheless be helpful to state it explicitly here

woodsp-ibm commented:
As its ready do you want to edit the [WIP} off the title and take it out of Draft

a-matsuo commented:
@woodsp-ibm Thank you for reviewing! I fixed points you mentioned and take the PR out of Draft.

woodsp-ibm commented on qiskit_finance/applications/optimization/portfolio_optimization.py:
```diff
@@ -0,0 +1,171 @@
+        """Getter of risk_factor
+
+        Returns:
+            The risk appetite of the decision maker.
+        """
+        return self._risk_factor
+
+    @risk_factor.setter
+    def risk_factor(self, risk_factor: float) -> None:
+        """Setter of risk_factor
+
+        Args:
+            risk_factor: The risk appetite of the decision maker.
```
Maybe the text should say `risk factor` I know people talk about whether you might have an appetite for risk but maybe it might not be so clear to some. Anyway its pretty minor since its on the setter/getter which should be more obvious so I approved anyway

stefan-woerner reviewed (approved):
lgtm
